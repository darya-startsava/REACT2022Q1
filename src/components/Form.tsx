import React from 'react';
import Input from './Input';
import DateInput from './DateInput';
import Select from './Select';
import RadioInput from './RadioInput';
import CheckboxInput from './CheckboxInput';
import FileInput from './FileInput';
import CardListFromFormPage from './CardListFromFormPage';
import CardFromFormPageType from '../types/card-from-form-page';

type State = {
  data: CardFromFormPageType[];
  successMessage: string;
  nameError: string;
  dateError: string;
  genderError: string;
  countryError: string;
  genresError: string;
  imageError: string;
  isSubmitButtonEnabled: boolean;
};

let gender = '';
let genres: Array<string> = [];
let fileToCreateUrl: File;

export default class Form extends React.Component<{}, State> {
  formRef: React.RefObject<HTMLFormElement>;
  inputRef: React.RefObject<HTMLInputElement>;
  dateInputRef: React.RefObject<HTMLInputElement>;
  selectRef: React.RefObject<HTMLSelectElement>;
  radioInputRef: React.RefObject<HTMLDivElement>;
  checkboxInputRef: React.RefObject<HTMLDivElement>;
  fileInputRef: React.RefObject<HTMLInputElement>;
  constructor(props: Record<string, unknown>) {
    super(props);
    this.state = {
      data: [],
      nameError: '',
      dateError: '',
      genderError: '',
      countryError: '',
      genresError: '',
      imageError: '',
      successMessage: '',
      isSubmitButtonEnabled: false,
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.validate = this.validate.bind(this);
    this.formRef = React.createRef();
    this.inputRef = React.createRef();
    this.dateInputRef = React.createRef();
    this.selectRef = React.createRef();
    this.radioInputRef = React.createRef();
    this.checkboxInputRef = React.createRef();
    this.fileInputRef = React.createRef();
  }

  handleChange() {
    if (this.state.successMessage) {
      this.setState({ successMessage: '' });
    }
    if (
      !this.state.nameError &&
      !this.state.dateError &&
      !this.state.genderError &&
      !this.state.countryError &&
      !this.state.genresError &&
      !this.state.imageError
    ) {
      this.setState({ isSubmitButtonEnabled: true });
    }
  }

  validate() {
    let nameError = '';
    let dateError = '';
    let countryError = '';
    let genderError = '';
    let genresError = '';
    let imageError = '';
    if (
      !this.inputRef.current?.value ||
      (this.inputRef.current?.value && this.inputRef.current?.value.length <= 1)
    ) {
      nameError = 'This field requires more than 1 symbol';
    }
    if (!this.dateInputRef.current?.value) {
      dateError = 'This field must be filled';
    }
    if (
      this.dateInputRef.current?.value &&
      Date.parse(this.dateInputRef.current!.value) >= Date.now()
    ) {
      dateError = 'Date must be earlier than today';
    }

    for (let i = 0; i < this.radioInputRef.current!.children.length; i++) {
      const input = this.radioInputRef.current?.children[i].children[0] as HTMLInputElement;
      if (input.checked) {
        gender = input.value;
      }
    }
    if (!gender) {
      genderError = 'Choose gender';
    }
    for (let i = 0; i < this.checkboxInputRef.current!.children.length; i++) {
      const input = this.checkboxInputRef.current?.children[i].children[0] as HTMLInputElement;
      if (input.checked) {
        genres.push(input.value);
      }
    }
    if (!this.selectRef.current?.value) {
      countryError = 'Choose country of birth';
    }
    if (!genres.length) {
      genresError = 'Choose at least one genre';
    }
    if (!this.fileInputRef.current?.files?.[0]) {
      imageError = 'Add image';
    }
    if (!imageError && !this.fileInputRef.current?.files?.[0].name.match(/.jpg$|.png$/)) {
      imageError = 'Add file with extension .jpg or .png';
    }
    if (nameError) {
      this.setState({ nameError });
    }
    if (dateError) {
      this.setState({ dateError });
    }
    if (genderError) {
      this.setState({ genderError });
    }
    if (countryError) {
      this.setState({ countryError });
    }
    if (genresError) {
      this.setState({ genresError });
    }
    if (imageError) {
      this.setState({ imageError });
    }
    if (nameError || dateError || genderError || countryError || genresError || imageError) {
      this.setState({ isSubmitButtonEnabled: false });
      return false;
    }
    this.setState({ nameError, dateError, genderError, genresError, countryError, imageError });
    this.setState({ isSubmitButtonEnabled: true });
    return true;
  }

  handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    gender = '';
    genres = [];
    event.preventDefault();
    if (this.state.successMessage) {
      this.setState({ successMessage: '' });
    }
    const isValid = this.validate();
    if (isValid) {
      if (this.fileInputRef.current?.files) {
        fileToCreateUrl = this.fileInputRef.current?.files[0];
      }
      this.setState(
        (state) => {
          return {
            data: state.data.concat({
              id: state.data.length,
              image: URL.createObjectURL(fileToCreateUrl),
              name: this.inputRef.current?.value,
              gender: gender,
              dateOfBirth: this.dateInputRef.current?.value,
              countryOfBirth: this.selectRef.current?.value,
              movieGenres: genres,
            }),
            successMessage: 'Your data has been successfully saved',
          };
        },
        () => {
          this.formRef.current?.reset();
        }
      );
    }
  }

  render() {
    return (
      <>
        <form onSubmit={this.handleSubmit} ref={this.formRef}>
          <Input
            ref={this.inputRef}
            onChange={() => this.setState({ nameError: '' }, this.handleChange)}
          />
          <br />
          <div style={{ fontSize: 12, color: 'red' }}>{this.state.nameError}</div>
          <RadioInput
            ref={this.radioInputRef}
            onChange={() => this.setState({ genderError: '' }, this.handleChange)}
          />
          <br />
          <div style={{ fontSize: 12, color: 'red' }}>{this.state.genderError}</div>
          <DateInput
            ref={this.dateInputRef}
            onChange={() => this.setState({ dateError: '' }, this.handleChange)}
          />
          <br />
          <div style={{ fontSize: 12, color: 'red' }}>{this.state.dateError}</div>
          <Select
            ref={this.selectRef}
            onChange={() => this.setState({ countryError: '' }, this.handleChange)}
          />
          <br />
          <div style={{ fontSize: 12, color: 'red' }}>{this.state.countryError}</div>
          <CheckboxInput
            ref={this.checkboxInputRef}
            onChange={() => this.setState({ genresError: '' }, this.handleChange)}
          />
          <br />
          <div style={{ fontSize: 12, color: 'red' }}>{this.state.genresError}</div>
          <FileInput
            ref={this.fileInputRef}
            onChange={() => this.setState({ imageError: '' }, this.handleChange)}
          />
          <br />
          <div style={{ fontSize: 12, color: 'red' }}>{this.state.imageError}</div>
          <button
            disabled={!this.state.isSubmitButtonEnabled}
            type="submit"
            className="btn btn-primary"
          >
            Submit
          </button>
          <div style={{ fontSize: 20, color: 'green' }}>{this.state.successMessage}</div>
        </form>
        <CardListFromFormPage data={this.state.data} />
      </>
    );
  }
}
