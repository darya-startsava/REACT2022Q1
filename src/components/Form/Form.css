.error-message {
  font-size: 12px;
  color: red;
}

.success-message {
  font-size: 20px;
  color: green;
}