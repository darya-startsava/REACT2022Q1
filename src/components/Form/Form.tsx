import React, { useState, useEffect } from 'react';
import { useForm } from 'react-hook-form';
import './Form.css';
import Input from '../Input';
import Select from '../Select';
import DateInput from '../DateInput';
import FileInput from '../FileInput';
import CardList from '../CardList';
import CardType from '../../types/card';
import FormValues from '../../types/formValues';
import { ErrorMessage } from '@hookform/error-message';
import movieGenres from '../../data/movie-genres';

interface FormProps extends React.HTMLProps<HTMLFormElement> {}

export default function Form(props: FormProps) {
  const [data, setData] = useState<CardType[]>([]);
  const [isNewCardCreatedNow, setIsNewCardCreatedNow] = useState<boolean>(false);
  const {
    register,
    formState: { errors, isSubmitSuccessful, isDirty },
    handleSubmit,
    reset,
  } = useForm<FormValues>();

  const onSubmit = (formData: FormValues) => {
    const movieGenres = formData?.movieGenres?.join(', ');
    setData((data) => [
      ...data,
      {
        ...formData,
        uploadedImage: URL.createObjectURL(formData.picture![0]),
        movieGenres,
        id: data.length,
        isFull: false,
      },
    ]);
    setIsNewCardCreatedNow(true);
  };

  useEffect(() => {
    if (isDirty) setIsNewCardCreatedNow(false);
  }, [isDirty]);

  useEffect(() => {
    reset({
      name: '',
      gender: '',
      dateOfBirth: '',
      countryOfBirth: '',
      movieGenres: [],
      picture: null,
    });
  }, [isSubmitSuccessful, reset]);

  return (
    <>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          {...register('name', {
            required: 'This field requires more than 1 symbol',
            minLength: { value: 2, message: 'This field requires more than 1 symbol' },
          })}
        />
        <div className="error-message mb-3">
          <ErrorMessage errors={errors} name="name" />
        </div>
        <div>
          Gender:
          <label className="mx-1">
            Male
            <input
              className="mx-1"
              type="radio"
              value="male"
              {...register('gender', {
                required: {
                  value: true,
                  message: 'Choose gender',
                },
              })}
            />
          </label>
          <label className="mx-1">
            Female
            <input
              className="mx-1"
              type="radio"
              value="female"
              {...register('gender', {
                required: {
                  value: true,
                  message: 'Choose gender',
                },
              })}
            />
          </label>
        </div>
        <div className="error-message mb-3">
          <ErrorMessage errors={errors} name="gender" />
        </div>
        <DateInput
          {...register('dateOfBirth', {
            required: 'This field must be filled',
            validate: (value) =>
              Date.parse(value) < Date.now() || 'Date must be earlier than today',
          })}
        />
        <div className="error-message mb-3">
          <ErrorMessage errors={errors} name="dateOfBirth" />
        </div>
        <Select {...register('countryOfBirth', { required: 'Choose country of birth' })} />
        <div className="error-message mb-3">
          <ErrorMessage errors={errors} name="countryOfBirth" />
        </div>
        <div className="mt-2">
          Movie genres:
          {movieGenres.map((item) => {
            return (
              <label key={item} className="mx-1">
                {item}
                <input
                  className="mx-1"
                  type="checkbox"
                  value={item}
                  {...register('movieGenres', {
                    required: {
                      value: true,
                      message: 'Choose at least one genre',
                    },
                  })}
                />
              </label>
            );
          })}
        </div>
        <div className="error-message mb-3">
          <ErrorMessage errors={errors} name="movieGenres" />
        </div>
        <FileInput
          {...register('picture', {
            validate: (value) =>
              (value && value?.[0].name.match(/.jpg$|.png$/) !== null) ||
              'Add file with extension .jpg or .png',
          })}
        />
        <div className="error-message mb-3">
          <ErrorMessage errors={errors} name="picture" />
        </div>
        <button
          disabled={
            !!errors.name ||
            !!errors.gender ||
            !!errors.dateOfBirth ||
            !!errors.countryOfBirth ||
            !!errors.movieGenres ||
            !!errors.picture ||
            !isDirty
          }
          type="submit"
          className="btn btn-primary"
        >
          Submit
        </button>
        {isNewCardCreatedNow && (
          <div className="success-message">Your data has been successfully saved</div>
        )}
      </form>
      <CardList data={data} />
    </>
  );
}
