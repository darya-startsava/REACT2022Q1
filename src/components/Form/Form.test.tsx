import Form from './Form';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('form', () => {
  it("should show error messages for empty fields, but shouldn't show them for completed fields(1-3)", () => {
    const { container } = render(<Form />);

    userEvent.type(getName(), 'Name');
    userEvent.click(getGender());
    userEvent.type(getDate(), '2021-04-04');
    userEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(container).toMatchSnapshot();
  });

  it("should show error messages for empty fields, but shouldn't show them for completed fields(4-6)", () => {
    const { container } = render(<Form />);
    const file = new File(['hello'], 'hello.png', { type: 'image/png' });

    userEvent.selectOptions(getCountry(), ['UK']);
    userEvent.click(getGenres());
    userEvent.click(screen.getByRole('button', { name: /Submit/i }));
    userEvent.upload(getImage(), file);
    expect(container).toMatchSnapshot();
  });

  it('should create a new card if all fields are filled', () => {
    const { container } = render(<Form />);

    completeForm();
    expect(container).toMatchSnapshot();
  });

  it('should create 3 cards if submit 3 times', () => {
    const { container } = render(<Form />);

    for (let i = 1; i <= 3; i++) {
      completeForm();
    }
    expect(container).toMatchSnapshot();
  });
});

function getName() {
  return screen.getByRole('textbox', {
    name: /name:/i,
  });
}

function getGender() {
  return screen.getByRole('radio', {
    name: /female/i,
  });
}

function getDate() {
  return screen.getByLabelText(/Date of Birth:/i);
}

function getCountry() {
  return screen.getByRole('combobox', { name: /Country of Birth:/i });
}

function getGenres() {
  return screen.getByRole('checkbox', {
    name: /science fiction/i,
  });
}

function getImage() {
  return screen.getByLabelText(/upload file/i);
}

function completeForm() {
  global.URL.createObjectURL = jest.fn();
  userEvent.type(getName(), 'Name');
  userEvent.click(getGender());
  userEvent.type(getDate(), '2021-04-04');
  const file = new File(['hello'], 'hello.png', { type: 'image/png' });
  userEvent.selectOptions(getCountry(), ['UK']);
  userEvent.click(getGenres());
  userEvent.upload(getImage(), file);
  userEvent.click(screen.getByRole('button', { name: /Submit/i }));
}
