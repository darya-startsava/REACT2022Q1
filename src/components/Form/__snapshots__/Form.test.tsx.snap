// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form should create 3 cards if submit 3 times 1`] = `
<div>
  <form>
    <label
      class="form-label"
    >
      Name:
      <input
        class="form-control"
        type="text"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <div>
      Gender:
      <label
        class="mx-1"
      >
        Male
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="male"
        />
      </label>
      <label
        class="mx-1"
      >
        Female
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="female"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    />
    <label
      class="form-label"
    >
      Date of Birth:
      <input
        class="form-control"
        type="date"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <label>
      Country of Birth:
      <select
        class="form-select"
      >
        <option
          value=""
        >
          --choose country--
        </option>
        <option
          value="USA"
        >
          USA
        </option>
        <option
          value="UK"
        >
          UK
        </option>
        <option
          value="Canada"
        >
          Canada
        </option>
        <option
          value="Australia"
        >
          Australia
        </option>
        <option
          value="France "
        >
          France 
        </option>
        <option
          value="Germany"
        >
          Germany
        </option>
        <option
          value="Sweden"
        >
          Sweden
        </option>
        <option
          value="Spain"
        >
          Spain
        </option>
        <option
          value="Denmark"
        >
          Denmark
        </option>
      </select>
    </label>
    <div
      class="error-message mb-3"
    />
    <div
      class="mt-2"
    >
      Movie genres:
      <label
        class="mx-1"
      >
        Science fiction
        <input
          class="mx-1"
          type="checkbox"
          value="science fiction"
        />
      </label>
      <label
        class="mx-1"
      >
        Biopic
        <input
          class="mx-1"
          type="checkbox"
          value="biopic"
        />
      </label>
      <label
        class="mx-1"
      >
        Disaster movie
        <input
          class="mx-1"
          type="checkbox"
          value="disaster movie"
        />
      </label>
      <label
        class="mx-1"
      >
        Fantasy
        <input
          class="mx-1"
          type="checkbox"
          value="fantasy"
        />
      </label>
      <label
        class="mx-1"
      >
        Period drama
        <input
          class="mx-1"
          type="checkbox"
          value="period drama"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    />
    <label
      class="mb-2"
    >
      Upload file:
      <input
        class="mx-1"
        type="File"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
    <div
      class="success-message"
    >
      Your data has been successfully saved
    </div>
  </form>
  <ul
    class="list-group list-group-horizontal container"
  >
    <li
      class="list-group-item col-3"
    >
      <div
        class="my-1"
      >
        <h5
          class="card-title"
        >
          Name
        </h5>
        <p>
          female
        </p>
        <p>
          Date of birth: 
          2021-04-04
        </p>
        <p>
          Country of birth: 
          UK
        </p>
        <p>
          Genres: 
          science fiction
        </p>
      </div>
    </li>
    <li
      class="list-group-item col-3"
    >
      <div
        class="my-1"
      >
        <h5
          class="card-title"
        >
          Name
        </h5>
        <p>
          female
        </p>
        <p>
          Date of birth: 
          2021-04-04
        </p>
        <p>
          Country of birth: 
          UK
        </p>
        <p>
          Genres: 
          science fiction
        </p>
      </div>
    </li>
    <li
      class="list-group-item col-3"
    >
      <div
        class="my-1"
      >
        <h5
          class="card-title"
        >
          Name
        </h5>
        <p>
          female
        </p>
        <p>
          Date of birth: 
          2021-04-04
        </p>
        <p>
          Country of birth: 
          UK
        </p>
        <p>
          Genres: 
          science fiction
        </p>
      </div>
    </li>
  </ul>
</div>
`;

exports[`form should create a new card if all fields are filled 1`] = `
<div>
  <form>
    <label
      class="form-label"
    >
      Name:
      <input
        class="form-control"
        type="text"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <div>
      Gender:
      <label
        class="mx-1"
      >
        Male
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="male"
        />
      </label>
      <label
        class="mx-1"
      >
        Female
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="female"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    />
    <label
      class="form-label"
    >
      Date of Birth:
      <input
        class="form-control"
        type="date"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <label>
      Country of Birth:
      <select
        class="form-select"
      >
        <option
          value=""
        >
          --choose country--
        </option>
        <option
          value="USA"
        >
          USA
        </option>
        <option
          value="UK"
        >
          UK
        </option>
        <option
          value="Canada"
        >
          Canada
        </option>
        <option
          value="Australia"
        >
          Australia
        </option>
        <option
          value="France "
        >
          France 
        </option>
        <option
          value="Germany"
        >
          Germany
        </option>
        <option
          value="Sweden"
        >
          Sweden
        </option>
        <option
          value="Spain"
        >
          Spain
        </option>
        <option
          value="Denmark"
        >
          Denmark
        </option>
      </select>
    </label>
    <div
      class="error-message mb-3"
    />
    <div
      class="mt-2"
    >
      Movie genres:
      <label
        class="mx-1"
      >
        Science fiction
        <input
          class="mx-1"
          type="checkbox"
          value="science fiction"
        />
      </label>
      <label
        class="mx-1"
      >
        Biopic
        <input
          class="mx-1"
          type="checkbox"
          value="biopic"
        />
      </label>
      <label
        class="mx-1"
      >
        Disaster movie
        <input
          class="mx-1"
          type="checkbox"
          value="disaster movie"
        />
      </label>
      <label
        class="mx-1"
      >
        Fantasy
        <input
          class="mx-1"
          type="checkbox"
          value="fantasy"
        />
      </label>
      <label
        class="mx-1"
      >
        Period drama
        <input
          class="mx-1"
          type="checkbox"
          value="period drama"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    />
    <label
      class="mb-2"
    >
      Upload file:
      <input
        class="mx-1"
        type="File"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
    <div
      class="success-message"
    >
      Your data has been successfully saved
    </div>
  </form>
  <ul
    class="list-group list-group-horizontal container"
  >
    <li
      class="list-group-item col-3"
    >
      <div
        class="my-1"
      >
        <h5
          class="card-title"
        >
          Name
        </h5>
        <p>
          female
        </p>
        <p>
          Date of birth: 
          2021-04-04
        </p>
        <p>
          Country of birth: 
          UK
        </p>
        <p>
          Genres: 
          science fiction
        </p>
      </div>
    </li>
  </ul>
</div>
`;

exports[`form should show error messages for empty fields, and shouldn't show them for completed fields(1-3) 1`] = `
<div>
  <form>
    <label
      class="form-label"
    >
      Name:
      <input
        class="form-control"
        type="text"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <div>
      Gender:
      <label
        class="mx-1"
      >
        Male
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="male"
        />
      </label>
      <label
        class="mx-1"
      >
        Female
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="female"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    />
    <label
      class="form-label"
    >
      Date of Birth:
      <input
        class="form-control"
        type="date"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <label>
      Country of Birth:
      <select
        class="form-select"
      >
        <option
          value=""
        >
          --choose country--
        </option>
        <option
          value="USA"
        >
          USA
        </option>
        <option
          value="UK"
        >
          UK
        </option>
        <option
          value="Canada"
        >
          Canada
        </option>
        <option
          value="Australia"
        >
          Australia
        </option>
        <option
          value="France "
        >
          France 
        </option>
        <option
          value="Germany"
        >
          Germany
        </option>
        <option
          value="Sweden"
        >
          Sweden
        </option>
        <option
          value="Spain"
        >
          Spain
        </option>
        <option
          value="Denmark"
        >
          Denmark
        </option>
      </select>
    </label>
    <div
      class="error-message mb-3"
    >
      Choose country of birth
    </div>
    <div
      class="mt-2"
    >
      Movie genres:
      <label
        class="mx-1"
      >
        Science fiction
        <input
          class="mx-1"
          type="checkbox"
          value="science fiction"
        />
      </label>
      <label
        class="mx-1"
      >
        Biopic
        <input
          class="mx-1"
          type="checkbox"
          value="biopic"
        />
      </label>
      <label
        class="mx-1"
      >
        Disaster movie
        <input
          class="mx-1"
          type="checkbox"
          value="disaster movie"
        />
      </label>
      <label
        class="mx-1"
      >
        Fantasy
        <input
          class="mx-1"
          type="checkbox"
          value="fantasy"
        />
      </label>
      <label
        class="mx-1"
      >
        Period drama
        <input
          class="mx-1"
          type="checkbox"
          value="period drama"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    >
      Choose at least one genre
    </div>
    <label
      class="mb-2"
    >
      Upload file:
      <input
        class="mx-1"
        type="File"
      />
    </label>
    <div
      class="error-message mb-3"
    >
      Add image
    </div>
    <button
      class="btn btn-primary"
      disabled=""
      type="submit"
    >
      Submit
    </button>
    <div
      class="success-message"
    />
  </form>
  <ul
    class="list-group list-group-horizontal container"
  />
</div>
`;

exports[`form should show error messages for empty fields, and shouldn't show them for completed fields(4-6) 1`] = `
<div>
  <form>
    <label
      class="form-label"
    >
      Name:
      <input
        class="form-control"
        type="text"
      />
    </label>
    <div
      class="error-message mb-3"
    >
      This field requires more than 1 symbol
    </div>
    <div>
      Gender:
      <label
        class="mx-1"
      >
        Male
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="male"
        />
      </label>
      <label
        class="mx-1"
      >
        Female
        <input
          class="mx-1"
          name="gender"
          type="radio"
          value="female"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    >
      Choose gender
    </div>
    <label
      class="form-label"
    >
      Date of Birth:
      <input
        class="form-control"
        type="date"
      />
    </label>
    <div
      class="error-message mb-3"
    >
      This field must be filled
    </div>
    <label>
      Country of Birth:
      <select
        class="form-select"
      >
        <option
          value=""
        >
          --choose country--
        </option>
        <option
          value="USA"
        >
          USA
        </option>
        <option
          value="UK"
        >
          UK
        </option>
        <option
          value="Canada"
        >
          Canada
        </option>
        <option
          value="Australia"
        >
          Australia
        </option>
        <option
          value="France "
        >
          France 
        </option>
        <option
          value="Germany"
        >
          Germany
        </option>
        <option
          value="Sweden"
        >
          Sweden
        </option>
        <option
          value="Spain"
        >
          Spain
        </option>
        <option
          value="Denmark"
        >
          Denmark
        </option>
      </select>
    </label>
    <div
      class="error-message mb-3"
    />
    <div
      class="mt-2"
    >
      Movie genres:
      <label
        class="mx-1"
      >
        Science fiction
        <input
          class="mx-1"
          type="checkbox"
          value="science fiction"
        />
      </label>
      <label
        class="mx-1"
      >
        Biopic
        <input
          class="mx-1"
          type="checkbox"
          value="biopic"
        />
      </label>
      <label
        class="mx-1"
      >
        Disaster movie
        <input
          class="mx-1"
          type="checkbox"
          value="disaster movie"
        />
      </label>
      <label
        class="mx-1"
      >
        Fantasy
        <input
          class="mx-1"
          type="checkbox"
          value="fantasy"
        />
      </label>
      <label
        class="mx-1"
      >
        Period drama
        <input
          class="mx-1"
          type="checkbox"
          value="period drama"
        />
      </label>
    </div>
    <div
      class="error-message mb-3"
    />
    <label
      class="mb-2"
    >
      Upload file:
      <input
        class="mx-1"
        type="File"
      />
    </label>
    <div
      class="error-message mb-3"
    />
    <button
      class="btn btn-primary"
      disabled=""
      type="submit"
    >
      Submit
    </button>
    <div
      class="success-message"
    />
  </form>
  <ul
    class="list-group list-group-horizontal container"
  />
</div>
`;
